File: base/templates/js/final.js
class deviceObject {
  constructor(id, label, layer) {
    this.data = {
      id: id,
      label: label,
      layer: layer
    },
      this.switchPorts = {
      },
      this.connected = 0;
  }
}

class portObject {
  constructor(source, target, sourceLabel, targetLabel) {
    this.data = {
      id: source.id + "-" + target.id,
      source: source.id,
      target: target.id
    }
    this.style = {
      sourceLabel: sourceLabel,
      targetLabel: targetLabel
    }
  }
}

var vlan = jsonData;
console.log(vlan);

/*Step 1: Calculate switches of each layers from input vlan, assign port for switches */

var swt_num = 0;
var sum_host = 0;
var numCount = 10;
vlan.forEach((value, index) => {
  sum_host += value.host;
  value.id = "vlan " + numCount.toString();
  numCount++;
});

console.log("Sum host: ", sum_host);
accSwt_num = Math.ceil((sum_host + 24) / 24);
console.log("access switches: ", accSwt_num);
distSwt_num = Math.ceil(accSwt_num / 2);   // distribution switch number
console.log("dist swithces: ", distSwt_num);

var port_require = sum_host + distSwt_num * accSwt_num;
var port_have = accSwt_num * 24;
if (port_have - port_require > 24) {
  accSwt_num -= 1;
  distSwt_num = Math.ceil(accSwt_num / 2);
}
console.log("port require: ", port_require);
console.log("port have: ", port_have);

for (var i = 0; i < vlan.length; i++) {
  // Check if the host count is more than 24
  if (vlan[i].host > (24 - distSwt_num)) {
    // Calculate the number of new VLANs needed to accommodate all hosts
    var numNewVlans = Math.ceil(vlan[i].host / (24 - distSwt_num));
    // Create new VLAN objects with 24 hosts each
    for (var j = 1; j <= numNewVlans; j++) {
      // Calculate the number of hosts for this new VLAN
      var newHostCount = Math.min(vlan[i].host - ((j - 1) * (24 - distSwt_num)), (24 - distSwt_num));
      // Create the new VLAN object and add it to the array
      vlan.push({
        id: vlan[i].id,
        host: newHostCount,
        name: vlan[i].name, port: [
          {
            switch: "",
            switchport: []
          }
        ]
      });
    }
    // Remove the original VLAN object from the array
    vlan.splice(i, 1);
    i--;
  }
}
vlan = vlan.sort((obj1, obj2) => obj2.host - obj1.host);
console.log(vlan)

var coreSwt_num = Math.ceil(distSwt_num / 3); //Calculate number of core switch
var coreDevice = [];
var distDevice = [];
var accessDevice = [];
var router = [];

function definePort(source, numPort, typePort) {
  counter = 1;
  octet = 0;
  for (let i = 1; i <= numPort; i++) // The number depend on number of ports which is fetched from database
  {
    if (counter > 24) {
      octet++;
      counter = 1;
    }
    if (typePort == "Gi") {
      portLabel = typePort + "1/" + octet + "/" + counter;
    } else if (typePort == "fa") {
      portLabel = typePort + octet + "/" + counter;
    }

    counter++;
    source.switchPorts[portLabel] = false;
  }
}
/*
Define Switch by layer in devices variable, switches must be defined before calculate port since 
we have to collect id of devices between 2 layers.
 */
for (let i = 0; i < accSwt_num; i++) {
  accessDevice.push(new deviceObject("accessSwt" + i, "Access Switch " + i, "access"));
  definePort(accessDevice[i], 24, "fa");
}
for (let i = 0; i < distSwt_num; i++) {
  distDevice.push(new deviceObject("distSwt" + i, "Distribution Switch " + i, "distribution"));
  definePort(distDevice[i], 24, "Gi");
}
for (let i = 0; i < coreSwt_num; i++) {
  coreDevice.push(new deviceObject("coreSwt" + i, "Core Switch " + i, "core"));
  definePort(coreDevice[i], 24, "Gi");
}
router.push(new deviceObject("router", "Router", "router"));
definePort(router[0], 2, "Gi");
/*-----------------------------------------------------*/



/* 
We prioritize upper layer than lower, so connection from upper layer is priority, then we work from core layer to distribution layer first 
*/

var portDevice = [];
for (let rt in router) {
  for (let core in coreDevice) {
    for (let portSource in router[rt].switchPorts) {
      if (router[rt].switchPorts[portSource] == false) {
        portTarget = Object.keys(coreDevice[core].switchPorts)[Object.keys(router[rt].switchPorts).indexOf(portSource)];
        if (coreDevice[core].switchPorts[portTarget] == false) {
          portDevice.push(new portObject(router[rt].data, coreDevice[core].data, portSource, portTarget));
          router[rt].switchPorts[portSource] = true;
          router[rt].connected++;
          coreDevice[core].switchPorts[portTarget] = true;
          coreDevice[core].connected++;
          break;
        }
      }
    }
  }
}
for (let core in coreDevice) {
  for (let dist in distDevice) {
    for (let portSource in coreDevice[core].switchPorts) {
      if (coreDevice[core].switchPorts[portSource] == false) {
        portTarget = Object.keys(distDevice[dist].switchPorts)[Object.keys(coreDevice[core].switchPorts).indexOf(portSource)];
        if (distDevice[dist].switchPorts[portTarget] == false) {
          portDevice.push(new portObject(coreDevice[core].data, distDevice[dist].data, portSource, portTarget));
          coreDevice[core].switchPorts[portSource] = true;
          coreDevice[core].connected++;
          distDevice[dist].switchPorts[portTarget] = true;
          distDevice[dist].connected++;
          break;
        }
      }
    }
  }
}

for (let dist in distDevice) {
  for (let access in accessDevice) {
    for (let portSource in distDevice[dist].switchPorts) {
      if (distDevice[dist].switchPorts[portSource] == false) {
        portTarget = Object.keys(accessDevice[access].switchPorts)[Object.keys(distDevice[dist].switchPorts).indexOf(portSource)];
        if (accessDevice[access].switchPorts[portTarget] == false) {
          portDevice.push(new portObject(distDevice[dist].data, accessDevice[access].data, portSource, portTarget));
          distDevice[dist].switchPorts[portSource] = true;
          distDevice[dist].connected++;
          accessDevice[access].switchPorts[portTarget] = true;
          accessDevice[access].connected++;
          break;
        }
      }
    }
  }
}

vlan.forEach((VLAN) => {
  check = false;
  for (let access in accessDevice) {
    if (VLAN.host <= 24 - accessDevice[access].connected) {
      for (let i = 0; i < VLAN.host; i++) {
        for (let port in accessDevice[access].switchPorts) {
          if (accessDevice[access].switchPorts[port] == false) {
            console.log(accessDevice[access].data.id);
            VLAN.port[0].switch = String(accessDevice[access].data.id);
            portIndex = Object.keys(accessDevice[access].switchPorts).indexOf(port);
            VLAN.port[0].switchport.push(portIndex);
            accessDevice[access].switchPorts[port] = true;
            accessDevice[access].connected++;
            break;
          }
        }
      }
      check = true;
    }
    if (check == true) {
      break;
    }
  }
});

vlan.forEach((VLAN, index) => {
  var labelTarget = "";
  var startPort = "";
  var endPort = "";
  for (let accdv in accessDevice) {
    if (VLAN.port[0].switch == accessDevice[accdv].data.id) {
      startPort = Object.keys(accessDevice[accdv].switchPorts)[VLAN.port[0].switchport[0]];
      for (let i = 0; i < VLAN.port[0].switchport.length; i++) {
        if (VLAN.port[0].switchport[i] + 1 == VLAN.port[0].switchport[i + 1]) {
          endPort = Object.keys(accessDevice[accdv].switchPorts)[VLAN.port[0].switchport[i + 1]];
        } else {
          if (endPort == "") {
            labelTarget = labelTarget + startPort + ",";
          } else {
            if (endPort.includes("fa") == true) {
              labelTarget = labelTarget + startPort + "-" + endPort.substring(4, endPort.length) + ",";
            } else if (endPort.includes("Gi") == true) {
              labelTarget = labelTarget + startPort + "-" + endPort.substring(6, endPort.length) + ",";
            }
          }
          startPort = Object.keys(accessDevice[accdv].switchPorts)[VLAN.port[0].switchport[i + 1]];
          endPort = "";
        }
      }

      labelTarget = labelTarget.substring(0, labelTarget.length - 1);

      portDevice.push(new portObject(accessDevice[accdv].data, VLAN, labelTarget, "Ethernet"));
    }
  }
})
console.log(vlan)
console.log(accessDevice);
console.log(portDevice);




//--------------------------------------------------------



/* CYTOSCAPE SECTION */



//--------------------------------------------------------




var cy = cytoscape({
  container: document.getElementById("cy"), // container to render in, declared in HTML

  wheelSensitivity: 0.1,

  style: [
    // the stylesheet for the graph
    {
      selector: "node",
      style: {
        "background-color": "#666",
        "label": "data(label)",
        "shape": "rectangle",
        "font-size": 10,
        "text-background-opacity": 1,
        "text-background-shape": "rectangle",
        "text-background-color": "black",
        "color": "white",
        "text-valign": "bottom"
      },
    },
    {
      selector: "edge",
      style: {
        "width": 3,
        "line-color": "#aaa",
        "curve-style": "bezier",
        "control-point-step-size": 80, // change this value to adjust the curve
        "line-style": "dashed",
        "line-dash-pattern": [6, 4],
        "font-size": 10,
        "text-wrap": "wrap",
        "text-max-width": "100px",
        "source-text-offset": "50px",
        "target-text-offset": "10px",
        "source-text-margin-y": "-10px",
        "target-text-margin-y": "-10px",
        "text-opacity": 0,
        "text-background-opacity": 1,
        "text-background-color": "black",
        "color": "yellow",
      },
    },
  ],
});

vlan.forEach(VLAN => {
  cy.add({
    data: { id: VLAN.id, label: VLAN.id },
    style: {
      "background-image": "./images/pc.png",
      "background-fit": "contain",
      "background-opacity": "0",
    }
  })
});

cy.add([{ data: router[0].data }]);

for (let i = 0; i < coreDevice.length; i++) {
  cy.add({
    data: coreDevice[i].data
  });
};

for (let i = 0; i < distDevice.length; i++) {
  cy.add({
    data: distDevice[i].data
  });
};

for (let i = 0; i < accessDevice.length; i++) {
  cy.add({
    data: accessDevice[i].data
  });
};

// cy.nodes('[layer="core"]').forEach(function (core) {
//   cy.add({
//     data: {
//       id: core.id() + "-" + "router", // unique id for the edge
//       source: "router", // set the source to the id of the "Router" node
//       target: core.id(), // set the target to the id of the current core node
//     },
//     style: {
//       "line-style": "solid",
//       "sourceLabel": "G0/0/1",
//       "targetLabel": "G1/0/11",
//     }
//   });
// });


for (let i = 0; i < portDevice.length; i++) {
  cy.add(
    {
      data: portDevice[i].data, style: portDevice[i].style
    },
  );
};

var edge = cy.getElementById('router-coreSwt0');

edge.style('sourceLabel', 'Gi0/0/1');

edge.style("line-style", "solid");

cy.nodes('[layer="core"]').style({
  "background-image": "./images/3650.png",
  "background-fit": "contain",
  "background-opacity": "0",
});

cy.nodes('[layer="distribution"]').style({
  "background-image": "./images/3650.png",
  "background-fit": "contain",
  "background-opacity": "0",
});

cy.nodes('[layer="access"]').style({
  "background-image": "./images/sw.png",
  "background-fit": "contain",
  "background-opacity": "0",
});

cy.nodes('[id="router"]').style({
  "background-image": "./images/router.png",
  "background-fit": "contain",
  "background-opacity": "0",
});

cy.on("select unselect", "edge", function (evt) {
  evt.target.style("text-opacity", evt.target.selected() ? 1 : 0);
  evt.target.style("line-color", evt.target.selected() ? "red" : "#aaa");
  evt.target.style("z-index", evt.target.selected() ? "3" : "1");
});

cy.on("tap", function (event) {
  if (event.target === cy) {
    cy.elements().style("line-color", "#aaa"); // reset line color of all edges
    cy.elements().style("z-index", "1");
  }
});

cy.nodes().on("tap", function (event) {
  var clickedNode = event.target;
  var connectedEdges = cy.elements().edgesWith(clickedNode);

  cy.elements().style("line-color", "#aaa"); // reset line color of all edges
  setTimeout(function () {
    connectedEdges.style("line-color", "black"); // highlight the edges connected to the clicked node
    connectedEdges.style("z-index", "3");
  }, 50);
});

var layout = cy.layout({
  name: "dagre",
  rankDir: "TB", // top to bottom
  rankSep: 150,
  nodeSep: 150,
});

layout.run();

File: cyto/final.py
from math import *
import jsonify
import json

class deviceObject:
    def __init__(self, id, label, layer):
        self.data = {"id": id, "label": label, "layer": layer}
        self.switchPorts = {}
        self.connected = 0
    def serialize(self):
        return {
            'data': self.data,
            'switchPorts': self.switchPorts,
            'connected': self.connected
        }

class portObject:
    def __init__(self, source, target, sourceLabel, targetLabel):
        self.data = {
            "id": source["id"] + "-" + target["id"],
            "source": source["id"],
            "target": target["id"],
        }
        self.style = {"sourceLabel": sourceLabel, "targetLabel": targetLabel}
    def serialize(self):
        return {
            'data': self.data,
            'style': self.style
        }

class Vlan:
    def __init__(self, host, id, name, switch, switchport):
        self.host = host
        self.id = id
        self.name = name
        self.port = {"switch": switch, "switchport": switchport}


# funtion to sort vlan ascending based on host value
def sortVlan(vlan):
    for x in range(0, len(vlan) - 1):
        for y in range(x + 1, len(vlan)):
            if vlan[x].host > vlan[y].host:
                vlan_tg = vlan[x]
                vlan[x] = vlan[y]
                vlan[y] = vlan_tg
    return vlan


def split_vlan(vlan):
    i = 0
    # Split vlan which has big value host
    while i < len(vlan):
        if vlan[i].host > (24 - ceil(len(vlan) / 2)):
            numSplit = ceil(
                (25 - len(vlan) - sqrt((25 - len(vlan)) ** 2 - 4 * vlan[i].host)) / 2
            )
            newHost = 24 - len(vlan) - numSplit + 1
            for j in range(numSplit - 1):
                newVlan = Vlan(newHost, vlan[i].id + "_" + str(j), vlan[i].name, "", [])
                vlan.append(newVlan)

            newVlan = Vlan(
                vlan[i].host - newHost * (numSplit - 1),
                vlan[i].id + "_" + str(numSplit - 1),
                vlan[i].name,
                "",
                [],
            )
            vlan.append(newVlan)
            vlan.pop(i)
        else:
            i = i + 1
    return vlan


# Merge vlan which has small value host
def mergeVlan(vlan):
    i = 0
    while i < len(vlan) - 1:
        if vlan[i].host + vlan[i + 1].host < 24 - ceil(len(vlan) / 2):
            newVlan = Vlan(
                vlan[i].host + vlan[i + 1].host,
                vlan[i].id + "+" + vlan[i + 1].id,
                vlan[i].name,
                "",
                [],
            )
            vlan.append(newVlan)
            vlan.pop(i)
            vlan.pop(i)
            sortVlan(vlan)
        else:
            i = i + 1
    return vlan


# function to define name and value port of deviceObject
def definePort(source, numPort, typePort):
    counter = 1
    octet = 0
    for i in range(numPort):
        if counter > 24:
            octet += 1
            counter = 1
        portLabel = ""
        if typePort == "Gi":
            portLabel = typePort + "1/" + str(octet) + "/" + str(counter)
        else:
            if typePort == "fa":
                portLabel = typePort + str(octet) + "/" + str(counter)
        counter += 1

        source.switchPorts[portLabel] = False


# Group and define Switch
def defineDevice(number, device_type, device_layer):
    devices = []
    for i in range(number):
        add_device = deviceObject(
            str(device_layer + device_type + str(i)),
            str(device_layer + " " + device_type + " ") + str(i),
            str(device_layer),
        )
        devices.append(add_device)
        if device_layer == "Core":
            definePort(devices[i], 24, "Gi")
        else:
            definePort(devices[i], 24, "fa")
    return devices


def createPortObject(port_array, source_devices, target_devices):
    count = 0
    for s_device in source_devices:
        for t_device in target_devices:
            for portSource in list(s_device.switchPorts.keys()):
                if s_device.switchPorts[portSource] == False:
                    portTarget = list(t_device.switchPorts.keys())[count]
                    if t_device.switchPorts[portTarget] == False:
                        new_port = portObject(
                            s_device.data, t_device.data, portSource, portTarget
                        )
                        port_array.append(new_port)
                        s_device.switchPorts[portSource] = True
                        s_device.connected += 1
                        t_device.switchPorts[portTarget] = True
                        t_device.connected += 1
                        break
        count += 1


# Connect 1 Vlan to 1 Access Switch
def connectVlan(vlan, accessDevice,distnum):
    for VLAN in vlan:
        portIndex = 0
        check = False
        for access in accessDevice:
            if access.connected==distnum:
                for i in range(VLAN.host):
                    for port in access.switchPorts:
                        if access.switchPorts[port] == False:
                            VLAN.port["switch"] = access.data["id"]
                            portIndex += 1
                            VLAN.port["switchport"].append(portIndex)
                            access.switchPorts[port] = True
                            access.connected += 1
                            break
                check = True
            if check == True:
                break


# Define port connect Access Switch to Vlan
def defineVlanPort(vlan, accessDevice, portDevice, number):
    for VLAN in vlan:
        labelTarget = ""
        startPort = ""
        endPort = ""
        for access in accessDevice:
            if VLAN.port["switch"] == access.data["id"]:
                startPort = list(access.switchPorts.keys())[number]
                endPort = list(access.switchPorts.keys())[number + VLAN.host - 1]
                if endPort.find("fa") != -1:
                    labelTarget = startPort + "-" + endPort[4:]
                if endPort.find("Gi") != -1:
                    labelTarget = startPort + "-" + endPort[6:]
                new_port = portObject(
                    access.data, {"id": VLAN.id}, labelTarget, "Ethernet"
                )
                portDevice.append(new_port)
def get_input_json():
    vlan_list = []
    file = open('./static/js/output.json', 'r')
    vlan = json.loads(file.read())
    for data in vlan:
        new_vlan = Vlan(int(data["host"]), data["name"],data["name"],"",[])
        vlan_list.append(new_vlan)
    return vlan_list
def cyto_json():
    # Data input here
    vlan0 = Vlan(70, "vlan0", "An", "", [])
    vlan1 = Vlan(1, "vlan1", "Anal", "", [])
    vlan2 = Vlan(15,"vlan2", "Analyst", "", [])
    vlan = get_input_json()
    # --------------
    sorted_vlan = sortVlan(vlan)
    splitted_vlan = split_vlan(sorted_vlan)
    merge_vlan = mergeVlan(splitted_vlan)
    # Caculated number of Switch, Dist, Core
    accSwt_num = len(vlan)
    distSwt_num = ceil(accSwt_num / 2)
    coreSwt_num = ceil(distSwt_num / 3)
    # ____________
    accessDevice = defineDevice(accSwt_num, "Switch", "Access")
    distDevice = defineDevice(distSwt_num, "Switch", "Dist")
    coreDevice = defineDevice(coreSwt_num, "Switch", "Core")
    # define Router port
    router = []
    add_router = deviceObject("router", "Router", "router")
    router.append(add_router)
    definePort(router[0], 2, "Gi")

    # Connect Router - Core
    portDevice = []
    createPortObject(portDevice, router, coreDevice)
    # Connect Core - Dist
    createPortObject(portDevice, coreDevice, distDevice)
    createPortObject(portDevice, distDevice, accessDevice)
    # Connect Dist - Access Switch
    connectVlan(merge_vlan, accessDevice, distSwt_num)
    defineVlanPort(vlan, accessDevice, portDevice, distSwt_num)
    networkDevice = accessDevice + distDevice + coreDevice + router
    networkDeviceJson = json.dumps([z.__dict__ for z in networkDevice])
    portDeviceJson = json.dumps([z.__dict__ for z in portDevice])
    vlanJson = json.dumps([z.__dict__ for z in merge_vlan])
    return networkDeviceJson, portDeviceJson, vlanJson
